docs/src/test/java/docs/http/javadsl/HttpServerDynamicRoutingExampleTest.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package docs.http.javadsl;
+
+import akka.NotUsed;
+import akka.actor.ActorSystem;
+import akka.http.javadsl.ConnectHttp;
+import akka.http.javadsl.Http;
+import akka.http.javadsl.marshallers.jackson.Jackson;
+import akka.http.javadsl.model.HttpRequest;
+import akka.http.javadsl.model.HttpResponse;
+import akka.http.javadsl.model.StatusCodes;
+import akka.http.javadsl.server.AllDirectives;
+import akka.http.javadsl.server.Route;
+import akka.stream.ActorMaterializer;
+import akka.stream.javadsl.Flow;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class HttpServerDynamicRoutingExampleTest extends AllDirectives {
+
+  public static void main(String[] args) throws Exception {
+    // boot up server using the route as defined below
+    ActorSystem system = ActorSystem.create("routes");
+
+    final Http http = Http.get(system);
+    final ActorMaterializer materializer = ActorMaterializer.create(system);
+
+    //In order to access all directives we need an instance where the routes are define.
+    HttpServerDynamicRoutingExampleTest app = new HttpServerDynamicRoutingExampleTest();
+
+    final Flow<HttpRequest, HttpResponse, NotUsed> routeFlow = app.createRoute().flow(system, materializer);
+
+    http.bindAndHandle(routeFlow, ConnectHttp.toHost("localhost", 8080), materializer);
+  }
+
+  //#dynamic-routing-example
+  private Map<String, Map<JsonNode, JsonNode>> state = new HashMap<>();
```

Comment:
This also needs to be made thread safe - `final private ... = new ConcurrentHashMap<>();`